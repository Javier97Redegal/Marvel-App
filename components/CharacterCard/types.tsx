type CharacterCardProps = {
    character: {
        id: number
        name: string | undefined
        thumbnail: {
            path: string
            extension: string
        }
    }
}