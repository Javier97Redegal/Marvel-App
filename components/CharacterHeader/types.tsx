export interface CharacterHeaderProps {
    id: number
}