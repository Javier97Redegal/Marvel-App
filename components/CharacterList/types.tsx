export interface CharacterListProps {
    mode: string
    query: string
}