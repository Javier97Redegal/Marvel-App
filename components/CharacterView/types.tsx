export interface CharacterViewProps {
    id: string
}