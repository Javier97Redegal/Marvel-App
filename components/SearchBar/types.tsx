import { TYPES_OF_CHARACTER_LIST } from '@/constants';

export interface SearchBarProps {
    type: TYPES_OF_CHARACTER_LIST
    results: number
}