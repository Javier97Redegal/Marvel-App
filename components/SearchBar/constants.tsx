export const INPUT_PLACEHOLDER = 'Write something!'