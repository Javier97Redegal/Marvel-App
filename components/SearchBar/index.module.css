.searchbar {
    max-width: 1980px;
    margin: 0 auto;
    padding: 3rem 2rem 0 2rem;
}

.searchbar small {
    display: block;
    text-transform: uppercase;
    font-size: 1.225rem;
    margin: 1rem 0 0 0;
}

.input {
    display: flex;
    gap: .5rem;
    padding: 0 0 .25rem 0;
    border-bottom: solid;
}

.input img {
    display: block;
    padding: .5rem;
    flex: 0 0 auto;
}

.input input {
    display: block;
    width: 100%;
    background: transparent;
    border: none;
    text-transform: uppercase;
    font-size: 1.5rem;
}

.input input:focus {
    outline: none;
}