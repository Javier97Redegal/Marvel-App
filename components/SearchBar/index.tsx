import { useEffect, useState } from 'react'
import { useSearch } from '@/context/SearchContext'
import { useFavorites } from '@/context/FavoritesContext'
import { INPUT_PLACEHOLDER } from './constants'
import { TYPES_OF_CHARACTER_LIST } from '@/constants'
import { SearchBarProps } from './types'

const SearchBar = ({ type }: SearchBarProps) => {
    const { searchTotal } = useSearch()
    const { searchFavorites } = useFavorites()
    const [inputValue, setInputValue] = useState<string>('')

    useEffect(() => {
        const delayInputTimeoutId = setTimeout(() => {
            switch (true) {
                case type === TYPES_OF_CHARACTER_LIST.FAVORITES:
                    searchFavorites(inputValue)

                    break

                default:
                    searchTotal(inputValue)

                    break
            }
        }, 500)
        return () => clearTimeout(delayInputTimeoutId)
    }, [inputValue, 500])

    const handleInputChange = (event: React.FormEvent<HTMLInputElement>) => {
        setInputValue(event.currentTarget.value)
    }

    return <>
        <input
            type='text'
            placeholder={INPUT_PLACEHOLDER}
            value={inputValue}
            onChange={handleInputChange}
        />
    </>
}

export default SearchBar