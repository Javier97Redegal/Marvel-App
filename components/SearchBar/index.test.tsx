import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import SearchBar from '@/components/SearchBar'
import { SearchBarProps } from '@/components/SearchBar/types'
import { useRouter } from 'next/navigation'
import { INPUT_PLACEHOLDER } from '@/components/SearchBar/constants'

jest.mock('next/navigation', () => ({
    useRouter: jest.fn()
}))


beforeAll(() => {
    jest.useFakeTimers()
})

afterAll(() => {
    jest.useRealTimers()
})

describe('SearchBar', () => {
    const mockProps: SearchBarProps = {
        mode: 'search',
        query: ''
    }

    const mockPush = jest.fn()
    const mockRouter = { push: mockPush }

    beforeEach(() => {
        (useRouter as jest.Mock).mockReturnValue(mockRouter)
    })

    it('renders search bar with initial state', () => {
        const { asFragment } = render(<SearchBar {...mockProps} />)

        expect(screen.getByPlaceholderText(INPUT_PLACEHOLDER)).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('updates input value on change', () => {
        render(<SearchBar {...mockProps} />)

        const input = screen.getByPlaceholderText(INPUT_PLACEHOLDER)
        fireEvent.change(input, { target: { value: 'Spider-Man' } })

        expect(input).toHaveValue('Spider-Man')
    })

    it('sets input value based on query prop', () => {
        const queryProps = { ...mockProps, query: 'Iron Man' }
        render(<SearchBar {...queryProps} />)

        const input = screen.getByPlaceholderText(INPUT_PLACEHOLDER)
        expect(input).toHaveValue('Iron Man')
    })

    it('calls router.push with correct query string after input change', () => {
        render(<SearchBar {...mockProps} />)

        const input = screen.getByPlaceholderText(INPUT_PLACEHOLDER)
        fireEvent.change(input, { target: { value: 'Hulk' } })

        expect(mockPush).not.toHaveBeenCalled()

        jest.advanceTimersByTime(500)

        expect(mockPush).toHaveBeenCalledWith('?mode=search&q=Hulk')
    })
})
