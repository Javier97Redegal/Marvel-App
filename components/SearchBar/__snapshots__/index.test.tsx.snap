// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders search bar with initial state 1`] = `
<DocumentFragment>
  <div
    class="searchbar"
  >
    <div
      class="input"
    >
      <img
        alt="Glass icon"
        src="/glass-icon.png"
      />
      <input
        placeholder="Search a character"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
